From 049cda48f6cb8bfbccf60f4f1423ccab5d6ebaee Mon Sep 17 00:00:00 2001
From: Haihao Xiang <haihao.xiang@intel.com>
Date: Fri, 10 Mar 2023 13:07:09 +0800
Subject: [PATCH 33/46] lavu/hwcontext_qsv: update AVQSVFramesContext to
 support dynamic frame pools

When AVQSVFramesContext.nb_surfaces is set to 0, a dynamic frame pool is
expected, otherwise a fixed frame pool is expected.

In addition, AVQSVFramesContext.info is added. User should provide
AVQSVFramesContext.info instead of AVQSVFramesContext.surfaces for
dynamic frame pools.

To reduce conflict in cartwheel, don't update libavutil/version.h and
doc/APIchanges !!!

Signed-off-by: Haihao Xiang <haihao.xiang@intel.com>
---
 libavutil/hwcontext_qsv.c |  4 ++--
 libavutil/hwcontext_qsv.h | 22 +++++++++++++++++++++-
 2 files changed, 23 insertions(+), 3 deletions(-)

diff --git a/libavutil/hwcontext_qsv.c b/libavutil/hwcontext_qsv.c
index 0f1ac50913..6d045405c4 100644
--- a/libavutil/hwcontext_qsv.c
+++ b/libavutil/hwcontext_qsv.c
@@ -613,7 +613,7 @@ static mfxStatus frame_alloc(mfxHDL pthis, mfxFrameAllocRequest *req,
     QSVFramesContext       *s = ctx->internal->priv;
     AVQSVFramesContext *hwctx = ctx->hwctx;
     mfxFrameInfo *i  = &req->Info;
-    mfxFrameInfo *i1 = &hwctx->surfaces[0].Info;
+    mfxFrameInfo *i1 = hwctx->nb_surfaces ? &hwctx->surfaces[0].Info : hwctx->info;
 
     if (!(req->Type & MFX_MEMTYPE_VIDEO_MEMORY_PROCESSOR_TARGET) ||
         !(req->Type & (MFX_MEMTYPE_FROM_VPPIN | MFX_MEMTYPE_FROM_VPPOUT)) ||
@@ -1173,7 +1173,7 @@ static int qsv_init_internal_session(AVHWFramesContext *ctx,
                               MFX_IOPATTERN_OUT_SYSTEM_MEMORY;
     par.AsyncDepth = 1;
 
-    par.vpp.In = frames_hwctx->surfaces[0].Info;
+    par.vpp.In = frames_hwctx->nb_surfaces ? frames_hwctx->surfaces[0].Info : *frames_hwctx->info;
 
     /* Apparently VPP requires the frame rate to be set to some value, otherwise
      * init will fail (probably for the framerate conversion filter). Since we
diff --git a/libavutil/hwcontext_qsv.h b/libavutil/hwcontext_qsv.h
index e2dba8ad83..317ae5094f 100644
--- a/libavutil/hwcontext_qsv.h
+++ b/libavutil/hwcontext_qsv.h
@@ -51,7 +51,27 @@ typedef struct AVQSVDeviceContext {
  * This struct is allocated as AVHWFramesContext.hwctx
  */
 typedef struct AVQSVFramesContext {
-    mfxFrameSurface1 *surfaces;
+    /**
+     * A pointer to mfxFrameSurface1 or mfxFrameInfo structure.
+     *
+     * When nb_surfaces is 0, it is a pointer to mfxFrameInfo structure,
+     * otherwise it is a pointer to mfxFrameSurface1.
+     */
+    union {
+        mfxFrameSurface1 *surfaces;
+        mfxFrameInfo     *info;
+    };
+
+    /**
+     * Number of frames
+     *
+     * A dynamic frame pool is required when nb_surfaces is 0, otherwise
+     * a fixed frame pool is required.
+     *
+     * User should make sure the configuration can support dynamic frame
+     * allocation when dynamic frame pool is required. For example, you cannt
+     * set nb_surfaces to 0 when the child_device_type is AV_HWDEVICE_TYPE_DXVA2.
+     */
     int            nb_surfaces;
 
     /**
-- 
2.34.1

